framework/Optimizers/GradientBasedOptimizer.py

```diff
@@ -265,86 +267,124 @@ def checkConvergence(self):
+    # if improved, keep it and move forward; otherwise, reject it and recommend cutting step size
+    if newerIsBetter:
+      self.status[traj]['reason'] = 'found new opt point'
+      self.raiseADebug('Accepting potential opt point for improved loss value')
+      #TODO REWORK this belongs in the base class optimizer; grad shouldn't know about multilevel!!
+      #  -> this parameter is how multilevel knows that a successful perturbation of an outer loop has been performed
+      #  maybe implement a "acceptPoint" method in base class?
+      self.mlActiveSpaceSteps[traj] += 1
+    else:
+      self.status[traj]['reason'] = 'rejecting bad opt point'
+      self.raiseADebug('Rejecting potential opt point for worse loss value. old: "{}", new: "{}"'.format(oldLossVal,currentLossVal))
+      # cut the next step size to hopefully stay in the valley instead of climb up the other side
+      self.recommendToGain[traj] = 'cut'
+
+    ## determine convergence
+    if pointFromRecommendation:
+      self.raiseAMessage('Setting convergence for Trajectory "{}" to "False" because of preconditioning.'.format(traj))
+      converged = False
+    else:
+      self.raiseAMessage('Checking convergence for Trajectory "{}":'.format(traj))
+      self.convergenceProgress[traj] = {} # tracks progress for grad norm, abs, rel tolerances
+      converged = False                   # updated for each individual criterion using "or" (pass one, pass all)
+      #printing utility
+      printString = '    {:<21}: {:<5}'
+      printVals = printString + ' (check: {:>+9.2e} < {:>+9.2e}, diff: {:>9.2e})'
+      def printProgress(name,boolCheck,test,gold):
+        """
+          Consolidates a commonly-used print statement to prevent errors and improve readability.
+          @ In, name, str, printed name of convergence check
+          @ In, boolCheck, bool, boolean convergence results for this check
+          @ In, test, float, value of check at current opt point
+          @ In, gold, float, convergence threshold value
+          @ Out, None
+        """
+        self.raiseAMessage(printVals.format(name,str(boolCheck),test,gold,abs(test-gold)))
+
+      # "min step size" and "gradient norm" are both always valid checks, whether rejecting or accepting new point
+
+      # min step size check
+      try:
+        lastStep = self.counter['lastStepSize'][traj]
+        minStepSizeCheck = lastStep <= self.minStepSize
+      except KeyError:
+        #we reset the step size, so we don't have a value anymore
+        lastStep = np.nan
+        minStepSizeCheck = False
+      printProgress('Min step size',minStepSizeCheck,lastStep,self.minStepSize)
+      converged = converged or minStepSizeCheck
+
+      # gradient norm
+      if len(self.counter['gradientHistory'][traj][0]) > 0: # that is, if we have gradient evaluations available ...
+        gradNorm = self.counter['gradNormHistory'][traj][0] # TODO the check used to be for len(...[1])>0
```

Comment:
My personal code of comment adding has generally been to include comments before if they are 1) multiline or 2) apply to several following lines of of code, and inline if it applies only to that line.  This is mostly to increase readability by allowing more consecutive lines of code to be adjacent without comments between them.  This might be worth bringing up as a standards question.

The comment in question is mostly a note to self to remind myself how it behaved before the rework of the gradient evaluation location.  I can remove the comment now.